### PR TITLE
Deny a patron with a root lane access to lanes and books not visible through that lane.

### PR DESCRIPTION
**Pull request by leonardr**
## Description

This branch removes a feature that was had not been used for a long time -- the library's loan policy -- and reimplements it in terms of an existing feature -- the patron's root lane.

A patron may or may not have a root lane. Before this branch, the root lane only controlled which lane a patron saw when they hit the root feed URL. Now, a patron's root lane puts additional restrictions on their activity:

* A patron with a root lane cannot view an OPDS feed for a Lanes that is 'above' their root lane 
* A patron with a root lane cannot view an OPDS feeds for a WorkList that might contain material not age-appropriate for their root lane. (As an example: a `ContributorLane` with `.audiences` set to include an age-inappropriate audience.)
* A patron with a root lane cannot access any controller for a book that is not age-appropriate for their root lane -- the circulation manager will send a problem detail. This includes the "permalink" controller but also the "borrow" and "place hold" controllers.
* A patron with a root lane cannot borrow a book that is not age-appropriate for their root lane. (But it shouldn't come to this because they won't be able to access the "borrow" controller in the first place.)

"Age-appropriate" is a concept defined in the [core branch](https://github.com/NYPL-Simplified/server_core/pull/1201) that goes along with this branch. In fact most of the code is in the core branch; this branch just uses that code to enforce the rules.

## Motivation and Context

https://jira.nypl.org/browse/SIMPLY-3125

## How Has This Been Tested?

`authenticated_patron_from_request` had no specific unit test coverage, so I added some. `load_lane` had test coverage but it was out of date because the class it was in had a name that didn't start with `Test`. I rewrote it.

In addition to the unit tests, I wrote some end-to-end tests illustrating the way people think of this feature: not in terms of certain methods being called, but in terms of certain patrons trying to take certain actions and not being allowed.

## Checklist:

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/adobe_vendor_id.py [hunk 1]
```diff
@@ -739,8 +739,11 @@ def from_config(cls, library, _db=None):
         ):
             raise CannotLoadConfiguration(
                 "Short Client Token configuration is incomplete. "
-                "vendor_id, username, password and "
-                "Library website_url must all be defined.")
+                "vendor_id (%s), username (%s), password (%s) and "
+                "Library website_url (%s) must all be defined." % (
+                    vendor_id, library_uri, library_short_name, secret
+                )
+            )
```
Comment (leonardr) [inline api/adobe_vendor_id.py:746]: This is an unrelated change that helps us recover more quickly from SIMPLY-2983 incidents -- with the current error message it's not clear which piece of configuration is missing.

File: api/controller.py [hunk 2]
```diff
@@ -625,9 +626,20 @@ def load_lane(self, lane_identifier):
             elif isinstance(lane, WorkList):
                 lane.children = [self._db.merge(child) for child in lane.children]
         else:
-            lane = get_one(
-                self._db, Lane, id=lane_identifier, library_id=library_id
-            )
+            try:
+                lane_identifier = int(lane_identifier)
+            except ValueError, e:
+                pass
```
Comment (leonardr) [inline api/controller.py:632]: Previously if you asked for a lane with a string name (rather than a numeric ID) you would get a database error.

File: api/controller.py [hunk 3]
```diff
@@ -625,9 +626,20 @@ def load_lane(self, lane_identifier):
...
+
+            if isinstance(lane_identifier, int):
+                lane = get_one(
+                    self._db, Lane, id=lane_identifier, library_id=library_id
+                )
+
+        if lane and not lane.visible_to(self.request_patron):
```
Comment (leonardr) [inline api/controller.py:639]: This is why I had to add `request_patron`. Authentication is not required for any controller that uses `load_lane`, so `authenticated_patron_for_request` probably hasn't been called yet. I need a yes-or-no answer as to whether there is an authenticated patron or not, and `authenticated_patron_for_request` can return all sorts of things.

File: api/controller.py [hunk 4]
```diff
@@ -694,9 +707,11 @@ def load_licensepooldelivery(self, pool, mechanism_id):
     def apply_borrowing_policy(self, patron, license_pool):
         if isinstance(patron, ProblemDetail):
             return patron
-        if not patron.can_borrow(license_pool.work, self.manager.lending_policy):
```
Comment (leonardr) [inline api/controller.py:697]: The `can_borrow` method and `lending_policy` have been removed.

File: api/controller.py [hunk 5]
```diff
@@ -1958,6 +1959,8 @@ class ProfileController(CirculationManagerController):
     def _controller(self):
         """Instantiate a CoreProfileController that actually does the work.
         """
+        # TODO: Probably better to use request_patron and check for
+        # None here.
```
Comment (leonardr) [inline api/controller.py:1963]: It's not a good idea to call `authenticated_patron_from_request` and assume it will return a `Patron`, but that's what's happening here. I don't want to spend time fixing it in this branch though.

File: api/routes.py [hunk 6]
```diff
@@ -81,6 +81,9 @@ def decorated(*args, **kwargs):
 def allows_auth(f):
     """Decorator function for a controller method that supports both
     authenticated and unauthenticated requests.
+
+    NOTE: This decorator might not be necessary; you can probably call
+    BaseCirculationManagerController.request_patron instead.
```
Comment (leonardr) [inline api/routes.py:86]: I experimented with marking the controllers that use `load_lane` as `@allow_auth` but I think it's actually better to get rid of the decorator.

File: api/controller.py [hunk 7]
```diff
@@ -255,9 +260,6 @@ def load_settings(self):
         self.circulation_apis = new_circulation_apis
         self.custom_index_views = new_custom_index_views
         self.shared_collection_api = self.setup_shared_collection()
-        self.lending_policy = load_lending_policy(
-            Configuration.policy('lending', {})
-        )
```
Comment (leonardr) [inline api/controller.py:260]: The lending policy was not tested or supported, and now has been removed.

File: tests/test_controller.py [hunk 8]
```diff
@@ -1199,52 +1297,72 @@ def test_library_for_request_reloads_settings_if_necessary(self):
...
             assert new_name in self.manager.auth.library_authenticators
 
+    def test_load_lane(self):
+        # Verify that requests for specific lanes are mapped to
+        # the appropriate lane.
 
-class FullLaneSetupTest(CirculationControllerTest):
-    """Most lane-based tests don't need the full multi-tier setup of lanes
-    that we would see in a real site. We use a smaller setup to save time
-    when running the test.
```
Comment (leonardr) [inline tests/test_controller.py:1206]: The test in this class was long out of date and not being run because its name didn't start with `Test`.

File: api/controller.py [hunk 9]
```diff
@@ -744,29 +761,20 @@ def authentication_document(self):
...
-        ).filter(
-            Lane.root_for_patron_type!=None
-        )
-        return root_lanes.count() > 0
+        """Does the active library feature root lanes for patrons of
+        certain types?
+
+        :return: A boolean
+        """
+        return flask.request.library.has_root_lanes
```
Comment (leonardr) [inline api/controller.py:769]: Implementation was moved to the `Library` class in core.

File: api/controller.py [hunk 10]
```diff
@@ -164,13 +163,26 @@ def __init__(self, _db, testing=False):
...
     def load_facets_from_request(self, *args, **kwargs):
-        """Load a faceting object from the incoming request, but also apply an authentication
-        restriction.
+        """Load a faceting object from the incoming request, but also apply some
+        application-specific access restrictions:
 
-        Specifically, in this application you can't use nonstandard
-        caching rules unless you're an authenticated administrator.
+        * You can't use nonstandard caching rules unless you're an authenticated administrator.
+        * You can't access a WorkList that's not accessible to you.
```
Comment (leonardr) [inline api/controller.py:170]: `load_facets_from_request` is called every time we're building an OPDS feed, and it already might short-circuit the request by returning a ProblemDetail (if the facets are incorrectly specified) so it's a convenient place to put other logic that might short-circuit the request by returning a ProblemDetail.

File: tests/test_controller.py [hunk 11]
```diff
@@ -1010,6 +1118,40 @@ def test_load_licensepools(self):
         )
         eq_(NO_LICENSES.uri, problem_detail.uri)
 
+    def test_load_work(self):
```
Comment (leonardr) [inline tests/test_controller.py:1121]: This method previously had no standalone tests.

File: tests/test_controller.py [hunk 12]
```diff
@@ -796,6 +829,80 @@ def test_unscoped_session(self):
...
+        o2 = object()
+        def set_patron():
+            flask.request.patron = o2
+        mock = MagicMock(side_effect = set_patron,
+                         return_value = "return value will be ignored")
+        self.controller.authenticated_patron_from_request = mock
+        with self.app.test_request_context("/"):
+            eq_(o2, self.controller.request_patron)
+
+    def test_authenticated_patron_from_request(self):
```
Comment (leonardr) [inline tests/test_controller.py:855]: A lot of this test coverage was missing.

File: api/controller.py [hunk 13]
```diff
@@ -647,7 +667,13 @@ def load_work(self, library, identifier_type, identifier):
 
         # We know there is at least one LicensePool, and all LicensePools
         # for an Identifier have the same Work.
-        return pools[0].work
+        work = pools[0].work
+
+        if work and not work.age_appropriate_for_patron(self.request_patron):
+            # This work is not age-appropriate for the authenticated
+            # patron. Don't show it.
+            work = NOT_AGE_APPROPRIATE
```
Comment (vbessonov) [inline api/controller.py:675]: Would it make sense to change the current approach and instead of returning `ProblemDetail` objects from functions and handling them with `if isinstance(result, ProblemDetail)` statements just raise a `ProblemDetail` exception handling it at the bottom of the callstack where we could log them appropriately and return a `ProblemDetail` object back to the caller if needed? Of course, I don't imply that it's needed to be done in this PR but maybe we can consider it for the future? I think it could simplify the codebase, we could get rid of a lot of `if isinstance(result, ProblemDetail)` statements

**Review by vbessonov — CHANGES_REQUESTED** [review]
This PR also looks good, added a couple of suggestions

File: api/controller.py (hunk 13 above)
Comment (leonardr) [reply api/controller.py:675]: I have had similar thoughts in the past and I think it could work. The main thing we'd have to change is situations where one piece of code calls `ProblemDetail.detailed` to add more information to a `ProblemDetail` object. I think on balance the code would be easier to understand if we made this change.
